### PR TITLE
First integration tests for provisioning api

### PR DESCRIPTION
**Pull request by DeepDiver1975**
@PVince81 @Raydiation @schiesbn @blizzz @karlitschek 

So - before touching the provisioning api to implement V2 - I wanted to have integration tests to be sure nothing will be broken.

TODO:
- setup jenkins on this branch
- integrate this into the regular ci mechanism (AFTER merge)
- test all provisioning APIs
- get coverage results

### CODE REVIEW

**Comment by karlitschek** [comment]
nice :+1:

**Comment by DeepDiver1975** [comment]
@MorrisJobke please have a look since you did use behat already as well - THX

**Comment by DeepDiver1975** [comment]
@cmonteroluque @rperezb just FYI - this is what I was talking about in the past regarding integration testing of the OCS API

File: build/integration/features/bootstrap/FeatureContext.php [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	public function usingApiVersion($version) {
+		$this->apiVersion = $version;
+	}
+
+	/**
+	 * @Given /^user "([^"]*)" exists$/
+	 */
+	public function userExists($user) {
+		throw new \Behat\Behat\Exception\PendingException();
+	}
```
Comment (LukasReschke) [inline build/integration/features/bootstrap/FeatureContext.php]: Indentation looks off compared with above.

Comment (DeepDiver1975) [inline build/integration/features/bootstrap/FeatureContext.php]: mimimi ... :see_no_evil:

File: build/integration/features/bootstrap/FeatureContext.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+// Require 3rd-party libraries here:
+//
+//   require_once 'PHPUnit/Autoload.php';
+//   require_once 'PHPUnit/Framework/Assert/Functions.php';
+//
+
+/**
+ * Features context.
+ */
+class FeatureContext extends BehatContext {
```
Comment (LukasReschke) [inline build/integration/features/bootstrap/FeatureContext.php]: Class name should be changed to something else

File: build/integration/composer.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+  "require-dev": {
+    "guzzlehttp/guzzle": "~5.0",
```
Comment (LukasReschke) [inline build/integration/composer.json]: Any reason that you used 5 vs. 6?
(https://github.com/guzzle/guzzle#version-guidance)

File: build/integration/composer.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+    "behat/behat": "2.4.*@stable"
```
Comment (LukasReschke) [inline build/integration/composer.json]: 3 rocks more and follows semantic versioning :speak_no_evil:

Comment (LukasReschke) [inline build/integration/composer.json]: See also http://docs.behat.org/en/v3.0/

**Comment by LukasReschke** [comment]
Besides me [doing mimimimimi](https://github.com/owncloud/core/pull/18105#discussion_r36415673), this is cool :rocket:

File: build/integration/composer.json (hunk 3 above)
Comment (DeepDiver1975) [inline build/integration/composer.json]: Well - I was reading the tutorial of 5 ...... :speak_no_evil:

File: build/integration/features/bootstrap/FeatureContext.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    /** @var string */
+    private $baseUrl = '';
+
+    /** @var ResponseInterface */
+    private $response = null;
```
Comment (Raydiation) [inline build/integration/features/bootstrap/FeatureContext.php]: Indention :hides:

File: build/integration/composer.json (hunk 4 above)
Comment (DeepDiver1975) [inline build/integration/composer.json]: me stupidly copying around stuff .... :hides:

File: build/integration/features/bootstrap/FeatureContext.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	/** @var int */
+	private $apiVersion = 1;
+
+    /**
+     * Initializes context.
+     * Every scenario gets it's own context object.
+     *
+     * @param array $parameters context parameters (set them up through behat.yml)
+     */
+    public function __construct(array $parameters) {
```
Comment (Raydiation) [inline build/integration/features/bootstrap/FeatureContext.php]: I suppose this constructor parameter is enforced by the test lib?

**Comment by DeepDiver1975** [comment]
@Raydiation are you by chance aware of a way to collect coverage data of php dev server?

**Comment by Raydiation** [comment]
@DeepDiver1975 what I've found: https://github.com/Behat/Behat/issues/92

**Comment by Raydiation** [comment]
TL;DR: Not needed for integration tests. 100% code coverage is only usable for unit tests because it would take too long otherwise

**Comment by Raydiation** [comment]
IMHO: write a simple parser for our API docs and run them against ownCloud :)

**Comment by Raydiation** [comment]
As in provide an example API call and the result, run the example call against the API with test data and then compare it :)

That way you generate API docs that are tested and also help you test stuff :)

**Comment by oparoz** [comment]
Or use Codeception which has an API tester

http://codeception.com/docs/10-WebServices

```
<?php
$I = new ApiTester($scenario);
$I->wantTo('create a user via API');
$I->amHttpAuthenticated('service_user', '123456');
$I->haveHttpHeader('Content-Type', 'application/x-www-form-urlencoded');
$I->sendPOST('/users', ['name' => 'davert', 'email' => 'davert@codeception.com']);
$I->seeResponseCodeIs(200);
$I->seeResponseIsJson();
$I->seeResponseContains('{"result":"ok"}');
?>
```

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 what I've found: Behat/Behat#92

I'm not talking about coverage on the behat process - but on the php -S localhost:8080

**Comment by DeepDiver1975** [comment]
> Or use Codeception which has an API tester

Well - there are plenty of tools - we finally simply need to get started ...

**Comment by Raydiation** [comment]
@DeepDiver1975 does unit test coverage fail because of the test server?

**Comment by Raydiation** [comment]
http://xdebug.org/docs/code_coverage btw

**Comment by oparoz** [comment]
> Well - there are plenty of tools - we finally simply need to get started ...

I have found that there were not that many, but I agree, tools should not slow down any effort to improve tests.

What I've gathered so far for Codeception

#### +

* Faster
* More readable tests
* Does Behavioural, functional and unit with one "language", not intermediary language needed
* Works better when lots of tests are needed

#### -

* New syntax to learn for die-hard PHPunit fans

**Comment by DeepDiver1975** [comment]
Jenkins is green :tada:

**Comment by DeepDiver1975** [comment]
> What I've gathered so far for Codeception

I was looking at codeception multiple times - I don't want to introduce another paradigm

**Comment by oparoz** [comment]
> I was looking at codeception multiple times - I don't want to introduce another paradigm

OK. Just know that, although I haven't tested with core tests yet, my PHPUnit tests in other projects run without any changes. You only need to familiarise yourself with the new syntax and setup if your intend on writing new kind of tests (behavioural, API).

**Comment by MorrisJobke** [comment]
> As in provide an example API call and the result, run the example call against the API with test data and then compare it :)
> 
> That way you generate API docs that are tested and also help you integration test stuff :)

@Raydiation I wrote exactly this some weeks back for another project ;) But it's not in the shape I want it

**Comment by DeepDiver1975** [comment]
So @oparoz wrote one test with codeception - https://github.com/owncloud/gallery/blob/b84508a0b687b91b623c6f81ea99e5fe40daa846/tests/api/CreateUserCept.php

Looks interesting as well .... which path to choose?

@MorrisJobke @PVince81 @icewind1991 @blizzz

**Comment by oparoz** [comment]
For the API stuff, Codeception is easy to use and coders can quickly get going. It's literally about typing a couple of commands and writing your first test, the same way you write a scenario. If you need to show the steps involved to a BA or a PO, then you can convert the script to English.

### Example
```
$I->amHttpAuthenticated('admin', 'admin');
$I->haveHttpHeader('Content-Type', 'application/x-www-form-urlencoded');
$I->sendPOST($baseUrl . '/users', ['userid' => 'test', 'password' => 'test']);
$I->seeResponseCodeIs(200);
$I->seeResponseIsXml();
$I->seeXmlResponseIncludes(
	XmlUtils::toXml(
		['status'   => 'ok']
));
```

becomes

```
* I am http authenticated "admin","admin"
* I have http header "Content-Type","application/x-www-form-urlencoded"
* I send post "/ocs/v1.php/cloud/users",{"userid":"test","password":"test"}
* I see response code is 200
* I see response is xml
* I see xml response includes "DOMDocument"
```
Of course, with Behat, once the commonly used steps are written in PHP, then things can go very fast as well from scenario to test and it might be more flexible.

### Some potential problems
* it will create a bunch of folders inside the folder you want to use, so, for core, you have to choose wisely. Maybe `build/integration`. The structure looks like this: https://github.com/owncloud/gallery/tree/testing-with-codeception/tests
* if you want to use the codeception built-in server (add some config lines and you have a server running during the tests only), then you'll need to use composer to install all the dependencies and run `codecept` from `vendor/bin`. Otherwise, you can just use the phar to generate and run stuff. There may be a way around that. I haven't dug too deep.
* if your QA engineers are used to working with Gherkin and have a backlog of acceptance tests, then you may not want to have 3 tests systems: PHPUnit for unit/integration, Behat for acceptance, Codeception for api/integration. In my projects, I prefer to use one tool for all three and it works because Codeception can execute PHPUnit tests.

I've written some more (more!?! ;)) on my decision to use it for Gallery here: https://github.com/owncloud/gallery/pull/264#issuecomment-128978681

**Comment by LukasReschke** [comment]
FWIW I do prefer Gherkin (Behat) syntax.

**Comment by oparoz** [comment]
It might be worth getting some feedback via the mailing list?

**Comment by oparoz** [comment]
An older comparison of the two from 2 years ago. Things might have changed on both sides
http://codeception.com/05-06-2013/specification-testing-coparison.html

**Comment by MorrisJobke** [comment]
I also like the gherkin syntax more.

**Comment by MorrisJobke** [comment]
Tested and the code looks good. :+1:

**Comment by DeepDiver1975** [comment]
I'm merging this now @SergioBertolinSG any future test case via pull request please - THX
